### DISCUSSION TITLE
Implementing Households Under Groups

### DISCUSSION THREAD

**Discussion by hay-kot** [mealie-recipes/mealie#2134] [Feature Request]
**Before submitting this feature request I have**
_COMPLETE AND FILL THESE OUT_
- [x] Reviewed the documentation and verified that my feature request isn't already a feature.
- [x] Reviewed the existing feature requests and my feature has not been requested.
- [x] Checked the [tasks taged](https://github.com/hay-kot/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

**Please Describe The Problem To Be Solved**

Since 

- https://github.com/hay-kot/mealie/issues/1326

Is not implemented, it's difficult to share recipes between households in any meaningful way. Even with #1326 implemented it only offers READ support without any tight integration into the rest of the application. Here's a better idea of what _I think_ people want when they say "Public Recipes"

_As a user I want to be able to share recipes with my entire extended family and friends in Mealie while maintaining a separate collection of Shopping Lists, Meal Plans, etc..._

Though some users may want to share recipes publicly with their friends/family, I think this may cover more use cases. 

**(Optional): Suggest A Solution**

I suggest the we implement another tier of organization underneath the `Group` type named `Household`

**General Thoughts**

- Shared Amongst Group
  - Recipes
  - Categories
  - Tags
  - Tools
  - Cookbooks
  - Foods

- Household Specific Items
  - Meal Plans
  - Web Hooks
  - Notifiers
  - Shopping Lists

--- 

Other Notes

- Some group settings would have to be moved to households like first day of the week
- Invitations would need to be updated to support households
- We'd need a new concept of Group Administrators who are able to create new households and invite users to those households
- We'd also need to upgrade our permissions model to offer some finer tuning of who can do what, especially with households. Something like an option to make a recipe "Read Only Outside of Group" or something - needs lots more fleshing out of ideas here. 

# This is still just an idea, please leave feedback and/or thoughts on if you'd want to see this or if it's going in the wrong direction.

**Comment by hay-kot** [comment]
From the discord

> Just as another viewpoint of how I'd use mealie with my family:
On the one hand having family that wants to use mealie for their own recipes too. They'd want their own list of recipes (because our tastes differ + we would edit/add notes to recipes differently) + viewing my recipes (a different group) if they want to try out something they ate at my home. They might want to add it to their own list if they like it.
Another case is family that doesn't want to use mealie and doesn't really want to log in/make an account (think of parents who aren't used to using a phone/computer and would mainly print out the recipes). They'd just want to browse my recipes via a link.
>
> Both cases could be fixed in various ways (e.g. making more accounts).
But for the first one, a share feature to share all recipes in your group would be the solution. This can be sharing with groups on the same instance (still needing to log in), sharing via a secret URL like how it is done with recipes, but then for a whole group, or a setting to make everything public.
The second case only works with a public recipe view of a group, be it public for everyone or with a secret URL.
>
> I don't think these ways of using mealie are supported with the above implementation.
>
> Hopefully that gives some insight of another way to use it :)

**Comment by rickitaly** [comment]
I would love it, it would cover also my request in https://github.com/hay-kot/mealie/discussions/2093

**Comment by kteppris** [comment]
I totally agree, this would be really need.

In my case, I use Mealie with another person, and we liked to share recipes, but maybe not all of them. Also, we like to plan our meals that we are having together each week, but I also like to plan my own week. This could be achieved with having multiple Households, and I agree with your Household Specific Items. But one thing to add maybe is to give the possibility to share recipes between certain Households or not. This could maybe be achieved by household specific recipe books and group wide recipe books.

I would love if that is implemented, that I could use Mealie together with all my family instead of just myself :)

**Comment by JTunX** [comment]
im new to mealie and see a lot of potential in it, my family is very food orientated and mealie will make sharing recipes with each other much easier. i recently stumble upon it and had to give it a go. set it up and created users for my family as well as adding a bunch of recipes we all would like with the idea we could each add them as we go. the only downside to it so far is that as we are all in separate households. we cant share the same library of recipes and have sperate meal plan or shopping lists.  having the ability to share the equipment, recipes, cook books but have the meal plans and shopping lists separate (as the favorite's tab is) would be the icing on the cake.

**Comment by djmega152** [comment]
I'd love to see this feature!  

My use case is that I have different groups of friends and co-workers.  I want to provide them access to my recipe DB but keep them separate from the primary group & its permissions.  I'd love to have the ability to:

- Publish a recipe to 1 or more groups.  
- Able to set defaults such as new recipes are auto assigned to selected groups.  
- Define that a group has read-only access, meaning they can't create new recipes

I'll definitely be following this thread and hope to see this capability in the future.

Thanks!

**Comment by hunterjm** [comment]
Thank's to @michael-genson (I'm assuming) for pointing this out to me on Discord. I've done some thinking about this over the past few days and think for most use cases, a group is for all intents and purposes a household in the current application. I don't really see a use case to add additional differentiation, especially after https://github.com/mealie-recipes/mealie/pull/2525 covered most of the "public" use cases already. Below are the approaches I'm thinking about and posted in Discord. After looking through the codebase briefly, it feels like Option 1 would be easiest to implement initially and require the least amount of changes. Option 2 and 3 are complementary (3 builds on top of 2). Honestly, the more I think about it - all 3 options can be probably be done.

Option 1:
In the left navbar, link out to other public groups. When authenticated and viewing another group, still show search and public cookbooks, but hide meal planner through tools.

Option 2:
Improve search within a group to also include public recipes from other groups potentially optional based on a search filter toggle.

Option 3:
Rework frontend routes to not hide recipes behind `/g/{slug}`. Instead of redirecting `/` to the authenticated user's group, the base route is the `RecipeExplorerPage` modified to include all group recipes AND public recipes. There is a new "Group" filter for advanced search that displays a selector for public groups with public recipes and defaulted to "Has Any" with all groups selected allowing filtering by the group who added the recipe.

Regardless of option:
When authenticated and viewing recipes, still allow a user to add to favorites, meal plan, and shopping list. Also allow authenticated users to comment based on the group setting for allowing comments which is currently not possible.

**Comment by PartemImperium** [comment]
The biggest issue I see with requiring groups/recipes to be set to public for other group's to access them is that means that they are accessible anonymously without an account. At least as far as I know there isnt a way to share them with certain users or groups without sharing them with anyone and everyone.

**Reply by hunterjm** [reply]
That makes sense. I'm still digging into how permissions work currently. When I go to the group page and see this:
<img width="440" alt="image" src="https://github.com/mealie-recipes/mealie/assets/747670/10e1f0d0-5832-48b6-9437-92ef3251e112">
It makes me think authenticated users. It almost feels like we need to distinguish between anonymous sharing and authenticated sharing. I personally don't care if my recipes are public, and providing authenticated users an easier way to navigate them in the UI and use them for meal planning, etc would be helpful. What about an additional group setting for "Allow anonymous access to public recipes"?

This is the application flow today:
![Mealie-Today drawio](https://github.com/mealie-recipes/mealie/assets/747670/d3dce5bc-f846-4d7e-b45b-66ed998ed85c)

If we added a group setting for anonymous access, it could look like this:
![Mealie-Later drawio](https://github.com/mealie-recipes/mealie/assets/747670/dcf47029-fe7f-4270-89a2-0df7d2ee8c1f)

Either way, my options above currently **only** apply to how things work today. I just want to make already public recipes easier to discover for authenticated users. Adding the additional flag and flow for anonymous access can be a later feature.

**Comment by PartemImperium** [comment]
Ya an additional group setting for allow anonymous would solve that. But I do agree the ability to actually use a recipe outside of your group in a meal plan/shopping cart/ favorite is of more importance. 

Of the 3 options I like 2 the best but all 3 could be useful in certain usecases.

**Comment by hunterjm** [comment]
See #3017 for an initial stab at a basic implementation.

**Comment by RefineryX** [comment]
I posted something similar to this last year https://github.com/mealie-recipes/mealie/discussions/2401#discussioncomment-5971333. Besides OIDC (which is see is coming 🚀), the efficient sharing of recipes is the only thing stopping me from going all.

**Wanted to add my two cents:**
1. I don't think sharing should rely solely on public recipes (none of my recipes will be public) and only accessible to users with an account
2. I don't think recipes should be 'duplicated'. If sharing, there should be only one recipe and if it gets updated, it will update for all users.

I think how @hay-kot envisions it would be perfect.
> As a user I want to be able to share recipes with my entire extended family and friends in Mealie while maintaining a separate collection of Shopping Lists, Meal Plans, etc...

### Example
Husband and wife have an account each. They are able to both add their own recipes within their accounts which are kept private. They also created a shared space for recipes they would like to try / cook together.

**Comment by Shredder5262** [comment]
I am brand new to mealie and am on board with this as well, I was disappointed that group sharing doesn't work like i thought it did... 
I would like to actually see a 3 tier system with groups

Admin - Keeper of the kingdom (Self explanatory) 

House group - Any and all recipes have visibility at this level to only household members with the option to share out to child groups. House has the option to 'hide' recipes contributed by the 'Extended Family' group.  

Extended family group - (child group) A communal place where members of 'House' can share to extended family and also extended family can contribute recipes; but do not have access to things like assets, creating categories, creating tags, Creating tools but can create recipes and cookbooks. This tier is necessary to preserve semi-exclusivity and protect mom's secret recipe for lasagna that's a family secret. 

Public group - (child group, least privileged access) Public access to recipes does not require a login but can see recipes that are shared from 'house' and 'extended family group'. Access to create anything is restricted and only recipes shared publicly are here.

**Reply by michael-genson** [reply]
This is what is currently planned with households. The current setup is a single home, and there is a public view. The only thing missing is in-between, i.e. the household

**Reply by Shredder5262** [reply]
Is there an ETA for implementation on this?

**Comment by boilermaker-ben** [comment]
Thank you @hay-kot for the hard work on Mealie! It's amazing!

**I'm also interested in adding another layer of access.** I was surprised to see that "public" recipes from one group are visible without authentication, but hidden from an authenticated users from groups. From my perspective, this would be the ideal interaction that takes place:

1. **All owned recipes can be marked as private (default), public, viewable to any registered user**, or viewable to one or more other groups
2. **Allow recipes where access is given to live in a server-wide "pool"** of shared recipes that is viewable in a left menu navigation option or through a filter in the "search" area
3. **Add a button to all recipes not created by the authenticated user to add recipe to their personal recipes collection**

Regarding point 1: Giving the option to make a recipe viewable to one or more other specific groups seems like it would be the toughest and isn't probably a critical feature to include initially, though it would be nice to have that kind of granularity in sharing
Regarding Point 3:  Adding a recipe as a "view-only" recipe would be appealing to be aware of recipe tweaks made in the future and avoid redundancy on a server, but it seems easier to implement as just a duplication of the recipe to a personally-owned version that can be edited and won't disappear if the original creator deletes it

I'm aware **Organizers** might be tricky to decide whether to make them continue to be a shared asset or not with this approach. Also, in the event of this kind of feature being implemented, a JSON field for the recipe's creator (and maybe the group of the creator) would be a good addition for attribution.

I wanted to add to the conversation and poke the feature request at the same time. Cheers!

**Comment by CtznSniiips** [comment]
+1. I just installed Mealie and this is the only feature I'm missing.

**Comment by uberguy16** [comment]
First, I set up Mealie two or three weeks ago and we love it. My wife was very reluctant to use it (I do most of the cooking except Christmas baking), but last night she tried putting in a couple of her recipes from her worn, messy, yellowed sheets of paper she had them on and she actually was surprised how easy it was to use and relatively intuitive. So YAY for wife acceptance!

But, I agree, the groups doesn't work like I thought it would. I read through the discussion here and I think the household addition would solve this gap that we see right now.

**Comment by michael-genson** [comment]
FYI this is being addressed in https://github.com/mealie-recipes/mealie/pull/3970. Not everything is being implemented all at once, but the groundwork is in that PR
